Source/WebKit/UIProcess/API/gtk/WebKitPrintOperation.cpp

```diff
@@ -393,6 +424,222 @@ static void webkitPrintOperationPrintPagesForFrame(WebKitPrintOperation* printOp
 #endif
 }
 
+static GRefPtr<GtkPrinter> findFilePrinter(GtkPrintSettings* settings)
+{
+    struct FindPrinterData {
+        GRefPtr<GtkPrinter> printer;
+    } data = { nullptr };
+
+    gtk_enumerate_printers([](GtkPrinter* printer, gpointer userData) -> gboolean {
+        auto& data = *static_cast<FindPrinterData*>(userData);
+        auto* backend = gtk_printer_get_backend(printer);
+        if (!g_strcmp0(G_OBJECT_TYPE_NAME(backend), "GtkPrintBackendFile")) {
+            data.printer = printer;
+            return TRUE;
+        }
+        return FALSE;
+    }, &data, nullptr, TRUE);
+
+    return data.printer;
+}
+
+static void webkitPrintOperationSendPagesToPrintPortal(WebKitPrintOperation* printOperation, WebFrameProxy* webFrame, GtkPrintSettings* printSettings, GtkPageSetup* pageSetup)
+{
+    auto* priv = printOperation->priv;
+
+    RELEASE_ASSERT(priv->portalProxy);
+    RELEASE_ASSERT(priv->preparePrintResponse && priv->preparePrintResponse->token());
+    RELEASE_ASSERT(priv->preparePrintResponse->result() == WEBKIT_PRINT_OPERATION_RESPONSE_PRINT);
+
+    const char* applicationName = g_get_application_name();
+    GUniquePtr<char> jobName(g_strdup_printf("%s job #%u", applicationName ? applicationName : "WebKit", ++jobNumber));
+    GRefPtr<GtkPrinter> filePrinter = findFilePrinter(printSettings);
+
+    // Print the page into a temporary file
+    GUniqueOutPtr<char> filename;
+    int fd = g_file_open_tmp("webkitgtkprintXXXXXX", &filename.outPtr(), nullptr);
+    close(fd);
```

Comment:
The tmpfile is never, and TTBOMK cannot ever be, deleted by WebKit itself.

The Print portal can take an arbitrary amount of time to print this file, and neither we nor the portal can know for sure when the printer is done with reading the file.

However, because it's a tmpfile, it'll be stored in memory, and it's automatically deleted on restart.

In other news, what this patch does is exactly what GTK does in GtkPrintOperation.

Reply:
It will live forever on systems where /tmp is not on tmpfs and there is no regular tmpfile cleaner, but if there's nothing we can do about that, then don't worry about it.